cocos/renderer/gfx-validator/DeviceValidator.cpp

```diff
@@ -0,0 +1,204 @@
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#include "base/CoreStd.h"
+
+#include "BufferValidator.h"
+#include "CommandBufferValidator.h"
+#include "DescriptorSetLayoutValidator.h"
+#include "DescriptorSetValidator.h"
+#include "DeviceValidator.h"
+#include "FramebufferValidator.h"
+#include "InputAssemblerValidator.h"
+#include "PipelineLayoutValidator.h"
+#include "PipelineStateValidator.h"
+#include "QueueValidator.h"
+#include "RenderPassValidator.h"
+#include "SamplerValidator.h"
+#include "ShaderValidator.h"
+#include "TextureValidator.h"
+
+namespace cc {
+namespace gfx {
+
+DeviceValidator *DeviceValidator::_instance = nullptr;
+
+DeviceValidator *DeviceValidator::getInstance() {
+    return DeviceValidator::_instance;
+}
+
+DeviceValidator::DeviceValidator(Device *device) : Agent(device) {
+    DeviceValidator::_instance = this;
+}
+
+DeviceValidator::~DeviceValidator() {
+    CC_SAFE_DELETE(_actor);
+    DeviceValidator::_instance = nullptr;
+}
+
+bool DeviceValidator::doInit(const DeviceInfo &info) {
+    if (!_actor->initialize(info)) {
+        return false;
+    }
+
+    _context                                     = _actor->getContext();
+    _API                                         = _actor->getGfxAPI();
+    _deviceName                                  = _actor->getDeviceName();
+    _queue                                       = CC_NEW(QueueValidator(_actor->getQueue()));
+    _cmdBuff                                     = CC_NEW(CommandBufferValidator(_actor->getCommandBuffer()));
+    ((CommandBufferValidator *)_cmdBuff)->_queue = _queue;
+    _renderer                                    = _actor->getRenderer();
+    _vendor                                      = _actor->getVendor();
+    _caps                                        = _actor->_caps;
+    memcpy(_features, _actor->_features, (uint)Feature::COUNT * sizeof(bool));
+
+    CC_LOG_INFO("Device validator enabled.");
+
+    return true;
+}
+
+void DeviceValidator::doDestroy() {
+    if (_cmdBuff) {
+        ((CommandBufferValidator *)_cmdBuff)->_actor = nullptr;
+        CC_DELETE(_cmdBuff);
+        _cmdBuff = nullptr;
+    }
+    if (_queue) {
+        ((QueueValidator *)_queue)->_actor = nullptr;
+        CC_DELETE(_queue);
+        _queue = nullptr;
+    }
+
+    _actor->destroy();
+}
+
+void DeviceValidator::resize(uint width, uint height) {
+    _actor->resize(width, height);
+}
+
+void DeviceValidator::acquire() {
+    _actor->acquire();
+}
+
+void DeviceValidator::present() {
+    _actor->present();
+}
+
+void DeviceValidator::setMultithreaded(bool multithreaded) {
+    _actor->setMultithreaded(multithreaded);
+}
+
+CommandBuffer *DeviceValidator::createCommandBuffer(const CommandBufferInfo &info, bool hasValidator) {
+    CommandBuffer *actor = _actor->createCommandBuffer(info, true);
```

Comment:
I haven't seen the use of the hasValidator parameter.

Reply:
Oops, that's a mistake. the parameter is for device agent, not device validator.